module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+
+        let bond = bond_opt.unwrap_or_default();
+        let bond: u64 = BorshDeserialize::deserialize(&mut bond.value())
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondFormat)?;
+
+        let minimum_bond = self
+            .minimum_attester_bond
+            .get_or_err(working_set)
+            .expect("The minimum bond should be set at genesis");
+
+        // We then have to check that the bond was greater than the minimum bond
+        if bond < minimum_bond {
+            return Err(AttesterIncentiveErrors::UserNotBonded);
+        }
+
+        Ok(())
+    }
+
+    fn check_transition(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        if let Some(curr_tx) = self
+            .chain_state
+            .historical_transitions
+            .get(&max_attested_height, working_set)
+        {
+            // We first need to compare the initial block hash to the previous post state root
+            if !curr_tx.compare_hashes(&attestation.da_block_hash, &attestation.post_state_root) {
+                // It is the right attestation, we have to compare the initial block hash to the
+                // previous post state root
+                // Slash the attester
+                return Err(self.slash_and_invalidate_attestation(
+                    attester,
+                    max_attested_height,
+                    SlashingReason::TransitionInvalid,
+                    working_set,
+                ));
+            }
+            Ok(CallResponse::default())
+        } else {
+            Err(self.slash_burn_reward(
+                attester,
+                Role::Attester,
+                SlashingReason::TransitionNotFound,
+                working_set,
+            ))
+        }
+    }
+
+    fn check_initial_hash(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // Normal state
+        if let Some(transition) = self
+            .chain_state
+            .historical_transitions
+            .get(&(max_attested_height - 1), working_set)
+        {
+            if transition.post_state_root() != attestation.initial_state_root {
+                // The initial root hashes don't match, just slash the attester
+                return Err(self.slash_burn_reward(
+                    attester,
+                    Role::Attester,
+                    SlashingReason::InvalidInitialHash,
+                    working_set,
+                ));
+            }
+        } else {
+            // Genesis state
```

Comment:
Are we sure that this must be the genesis state? It seems safer to explicitly check that the transition number is 0. That way, we don't introduce a hidden invariant that the argument to this function must be less than the current block height. 

Alternatively, we can use a newtype to make the invariant explicit

Reply:
Added a `genesis_height` to the chain state module so that we can check the transition against that height.

Reply:
That prevents us from using an extra new type to make the invariant explicit.